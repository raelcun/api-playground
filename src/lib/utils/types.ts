import * as t from 'io-ts'

interface NonEmptyStringBrand {
  readonly NonEmptyString: unique symbol
}

export const NonEmptyString = t.brand(
  t.string,
  (s): s is t.Branded<string, NonEmptyStringBrand> => s !== undefined && s.length > 0,
  'NonEmptyString',
)

export type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>
}

export type ChangeTypeOfKeys<T, Keys extends keyof T, NewType> = {
  [key in keyof T]: key extends Keys ? NewType : T[key]
}

export type ArgumentTypes<T> = T extends (...args: infer U) => infer R ? U : never

export type ReplaceReturnType<T, TNewReturn> = (...a: ArgumentTypes<T>) => TNewReturn
