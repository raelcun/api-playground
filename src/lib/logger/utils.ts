import { either as E, taskEither as TE } from 'fp-ts'

import { Err } from '@lib/error'

import { Logger, LogMethods } from './types'

export const logErrorsE = <L extends Err<string>>(logger: Logger, method: LogMethods = 'trace') =>
  E.mapLeft<L, L>(e => {
    logger.log({ level: method, payload: e })
    return e
  })

export const logErrorsTE = <L extends Err<string>>(logger: Logger, method: LogMethods = 'trace') =>
  TE.mapLeft<L, L>(e => {
    logger.log({ level: method, payload: e })
    return e
  })
