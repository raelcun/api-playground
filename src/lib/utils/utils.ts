import { array as A, either as E, io as IO, taskEither as TE } from 'fp-ts'
import { pipe } from 'fp-ts/lib/pipeable'
import * as t from 'io-ts'
import { reporter } from 'io-ts-reporters'
import { Context, Middleware } from 'koa'

import { Err } from '@lib/error'

export const tryCatchTE: typeof TE.tryCatch = (f, onRejected) =>
  TE.tryCatch(() => {
    try {
      return f()
    } catch (err) {
      return Promise.reject(err)
    }
  }, onRejected)
tryCatchTE.call
export const createVoidTE = <L>() => TE.rightTask<L, void>(async () => {})

export const filterObjectKeys = <T extends Record<string, unknown>, K extends keyof T>(
  o: T,
  keys: K[],
): Pick<T, K> => {
  const res: any = {}

  keys.forEach(key => {
    res[key] = o[key]
  })

  return res
}

export const mapErrorCode = <T extends string>(newErrorCode: T) =>
  E.mapLeft((originalError: Err<string>) => ({
    ...originalError,
    code: newErrorCode,
    subcode: originalError.code,
  }))

export const decode = <T>(
  t: t.Type<T, unknown>,
  v: unknown,
): E.Either<Err<'VALIDATION_ERROR'>, T> =>
  pipe(
    t.decode(v),
    E.mapLeft(errors => ({
      code: 'VALIDATION_ERROR',
      message: reporter(E.left(errors)).join('. '),
    })),
  )

export const ioSequence = A.array.sequence(IO.io)

export const DateFromString = new t.Type<Date, string, unknown>(
  'DateFromString',
  (u): u is Date => u instanceof Date,
  (u, c) =>
    pipe(
      t.union([t.string, t.number]).validate(u, c),
      E.chain(s => {
        const d = new Date(s)
        return isNaN(d.getTime()) ? t.failure(u, c) : t.success(d)
      }),
    ),
  a => a.toISOString(),
)

export const createMiddlewareTE = <T>(
  f: (ctx: Context) => TE.TaskEither<unknown, unknown>,
): Middleware<T> => async (ctx, next) => {
  await pipe(
    f(ctx),
    TE.chain(() => TE.rightTask(next)),
  )()
}

export const createMiddlewareE = <T>(
  f: (ctx: Context) => E.Either<unknown, unknown>,
): Middleware<T> => createMiddlewareTE(ctx => TE.fromEither(f(ctx)))
