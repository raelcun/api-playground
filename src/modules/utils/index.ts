import { KoaContext } from '@modules/api-core'
import { createMockContext, Options } from '@shopify/jest-koa-mocks'

export const createKoaContext = <T>(
  options?: Options<Record<string, unknown>, T> & { requestBody?: T } & {
    body?: string | Record<string, unknown>
  },
): KoaContext<T> => {
  const context = createMockContext(options) as KoaContext<T>
  if (options !== undefined && options.body !== undefined) {
    context.body = options.body
  }
  return context
}

export const createMockNext = (impl = () => {}): jest.Mock<Promise<void>> =>
  jest.fn(() => {
    impl()
    return Promise.resolve()
  })
