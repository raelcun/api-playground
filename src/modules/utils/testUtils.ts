import * as HttpStatus from 'http-status-codes'
import { Context } from 'koa'

import { install, InstalledClock } from '@sinonjs/fake-timers'

export const testUnauthorized = (mockContext: Context, mockNext: jest.Mock<Promise<void>, []>) => {
  test('should set status to 401', () => {
    expect(mockContext.status).toEqual(HttpStatus.UNAUTHORIZED)
  })

  test('should not invoke next', () => {
    expect(mockNext).not.toHaveBeenCalled()
  })
}

export const testAuthorized = (mockContext: Context, mockNext: jest.Mock<Promise<void>, []>) => {
  test('should not set status to 401', () => {
    expect(mockContext.status).not.toEqual(HttpStatus.UNAUTHORIZED)
  })

  test('should invoke next once', () => {
    expect(mockNext).toHaveBeenCalledTimes(1)
  })
}

/* istanbul ignore next */
export const installStaticClock = () => {
  let clock: InstalledClock

  beforeAll(() => {
    clock = install({ now: 946684800000 })
  })

  afterAll(() => {
    clock.uninstall()
  })

  return () => clock
}
