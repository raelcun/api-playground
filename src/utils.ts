import Router from 'koa-router'
import Koa from 'koa'
import * as t from 'io-ts'

export const createServer = (router: Router): Koa =>
  new Koa().use(router.routes()).use(router.allowedMethods())

export const DateFromString = new t.Type<Date, string, unknown>(
  'DateFromString',
  (u): u is Date => u instanceof Date,
  (u, c) =>
    t.string.validate(u, c).chain(s => {
      const d = new Date(s)
      return isNaN(d.getTime()) ? t.failure(u, c) : t.success(d)
    }),
  a => a.toISOString(),
)
