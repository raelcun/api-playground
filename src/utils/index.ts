export * from './utils'
export * from './testUtils'
