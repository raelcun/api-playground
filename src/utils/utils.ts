import { array as A, either as E, io as IO, taskEither as TE } from 'fp-ts'
import { pipe } from 'fp-ts/lib/pipeable'
import t from 'io-ts'
import { reporter } from 'io-ts-reporters'
import { Context, Middleware } from 'koa'

import { Err } from '@modules/error/types'
import { Logger, LogMethods } from '@modules/logger/types'
import { KoaContext } from '@root/types'
import { createMockContext } from '@shopify/jest-koa-mocks'
import { Options } from '@shopify/jest-koa-mocks/src/create-mock-context'

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const createVoidTE = <L>() => TE.rightTask<L, void>(async () => {})

export const filterObjectKeys = <T extends object, K extends keyof T>(
  o: T,
  keys: K[],
): Pick<T, K> => {
  const res: any = {}

  keys.forEach(key => {
    res[key] = o[key]
  })

  return res
}

export const mapErrorCode = <T extends string>(newErrorCode: T) =>
  E.mapLeft((originalError: Err<string>) => ({
    ...originalError,
    code: newErrorCode,
  }))

export const logErrors = <L extends Err<string>>(logger: Logger, method: LogMethods = 'trace') =>
  E.mapLeft<L, L>(e => {
    logger[method](e)
    return e
  })

export const decode = <T>(
  t: t.Type<T, unknown>,
  v: unknown,
): E.Either<Err<'VALIDATION_ERROR'>, T> =>
  pipe(
    t.decode(v),
    E.mapLeft(errors => ({
      code: 'VALIDATION_ERROR',
      message: reporter(E.left(errors)).join(' '),
    })),
  )

export const ioSequence = A.array.sequence(IO.io)

export const createKoaContext = <T>(
  options?: Options<object, T> & { requestBody?: T } & { body?: string | object },
): KoaContext<T> => {
  const context = createMockContext(options) as KoaContext<T>
  if (options !== undefined && options.body !== undefined) {
    context.body = options.body
  }
  return context
}

export const createMockNext = () => jest.fn(() => Promise.resolve())

export const DateFromString = new t.Type<Date, string, unknown>(
  'DateFromString',
  (u): u is Date => u instanceof Date,
  (u, c) =>
    pipe(
      t.union([t.string, t.number]).validate(u, c),
      E.chain(s => {
        const d = new Date(s)
        return isNaN(d.getTime()) ? t.failure(u, c) : t.success(d)
      }),
    ),
  a => a.toISOString(),
)

export const createMiddlewareTE = <T>(
  f: (ctx: Context) => TE.TaskEither<unknown, unknown>,
): Middleware<T> => async (ctx, next) => {
  await pipe(
    f(ctx),
    TE.chain(() => TE.rightTask(next)),
  )()
}

export const createMiddlewareE = <T>(
  f: (ctx: Context) => E.Either<unknown, unknown>,
): Middleware<T> => createMiddlewareTE(ctx => TE.fromEither(f(ctx)))
