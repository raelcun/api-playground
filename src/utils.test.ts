import { either as E, taskEither as TE } from 'fp-ts'
import { DateFromString, createMiddlewareTE, createKoaContext, createMockNext } from './utils'

describe('DateFromString', () => {
  test.each<string | number>([
    '2019/1/1',
    '2019-1-1',
    '2019-10-10T14:48:00',
    'Wed, 09 Aug 1995 00:00:00 GMT',
    '2019-10-10T14:48:00.000+09:00',
    1565484560,
  ])('should successfully validate %s', date => {
    expect(E.isRight(DateFromString.decode(date))).toBe(true)
  })

  test.each<unknown>(['foo-bar', '23/25/2014', null, undefined])(
    'should fail to validate %s',
    date => {
      expect(DateFromString.is(date)).toBe(false)
      expect(E.isLeft(DateFromString.decode(date))).toBe(true)
    },
  )

  test('should successfully encode', () => {
    expect(DateFromString.encode(new Date('2019/1/1'))).toBe('2019-01-01T05:00:00.000Z')
  })
})

describe('createMiddlewareTE', () => {
  test('should call next on right', async () => {
    const mockNext = createMockNext()
    await createMiddlewareTE(() => TE.right('foo'))(createKoaContext(), mockNext)
    expect(mockNext).toHaveBeenCalledTimes(1)
  })

  test('should not call next on left', async () => {
    const mockNext = createMockNext()
    await createMiddlewareTE(() => TE.left('foo'))(createKoaContext(), mockNext)
    expect(mockNext).toHaveBeenCalledTimes(0)
  })
})
