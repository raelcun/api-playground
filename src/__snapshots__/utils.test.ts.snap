// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createKoaContext should wrap createMockContext 1`] = `
Array [
  Array [
    Object {
      "requestBody": "foo",
    },
  ],
]
`;

exports[`decode should report all errors joined together 1`] = `
Object {
  "_tag": "Left",
  "left": Object {
    "code": "VALIDATION_ERROR",
    "message": "Expecting number at foo but instead got: \\"foo\\". Expecting string at bar but instead got: 5.",
  },
}
`;

exports[`decode should return validation error when invalid input 1`] = `
Object {
  "_tag": "Left",
  "left": Object {
    "code": "VALIDATION_ERROR",
    "message": "Expecting string but instead got: 5.",
  },
}
`;

exports[`logErrors should invoke error method on logger with error 1`] = `
Array [
  Array [
    Object {
      "code": "foo",
    },
  ],
]
`;

exports[`logErrors should invoke info method on logger with error 1`] = `
Array [
  Array [
    Object {
      "code": "foo",
    },
  ],
]
`;

exports[`logErrors should invoke trace method on logger with error 1`] = `
Array [
  Array [
    Object {
      "code": "foo",
    },
  ],
]
`;

exports[`mapErrorCode should return initial error with updated error code 1`] = `
Object {
  "_tag": "Left",
  "left": Object {
    "code": "newCode",
    "message": "old message",
  },
}
`;
